feat: add dependency injections

DarrenSeng opened the pull request:
**Description**

This Draft PR addresses Issue #543 
- Router level dependency injections for sync & async handler functions
- Subrouters, their dependencies, and subrouters inheriting dependencies from their greater Router.
   ● Potentially addressed collisions between subrouter vs greater Router dependencies that share the same name
- Handler functions able to have parameters be unordered

Specific locations for changes: 
- Modified functions init.py: _add_route(), init(), include_router()
- Created functions in init.py: inject(), get_injected_dependencies()
- Modified functions in router.py async_inner_handler and inner_handler to inspect handler function parameters. 
● Logic to deal with unordered params: 'request' is now mapped to Request object inside of dependency dictionary, In inner_handler, you analyze the current handler function's parameters and filter out any dependency found in the dependencies dictionary vs any parameter that isn't a specified dependency in the dependencies dictionary. All dependencies found in the parameters get added to depToPass, including request, which depToPass is used in the formatted response that inner_handler returns. 
● Returns error if the handler function parameters contains a parameter that isn't a specified dependency in the dependencies dictionary accesses

sansyrox reviewed (commented):
Hey @DarrenSeng 👋 

Great work so far 😄  

I will do a thorough review soon but I have one suggestion right now. Do let me know if you require any more information 😄

sansyrox commented on main.py:
```diff
@@ -0,0 +1,63 @@
+        return "working example 3 returned"
+
+mainApp.inject("/s",example1=ex1) #specified inject, if you set to =ex1(), it will point to int return object
+mainApp.inject(example2=ex2) #unspecified inject
```
The inject function should allow multiple dependency injections in one call.

e.g. `mainApp.inject(example1=ex1, example2=ex2, ...)`

DarrenSeng replied:
Hello @sansyrox, just addressed that in the most recent commit

koryca replied:
Hello @sansyrox Would you mind providing some unit test cases to multi-route testing?

sansyrox commented on robyn/__init__.py:
```diff
@@ -31,8 +31,9 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
+        self.testName = ""
         self.config = config
-
+        self.dependencies = {"all":{'request':Request, 'response':Response}} #new
```
Can you please explain the logic behind the structure here?

DarrenSeng replied:
In inner_handler(..), you scan paramList and compare to dependencies dictionary for unknown dependency arguments. Request is a valid non-dependency parameter, so there needed to be a way to account for that when checking, which is why 'request' is mapped to a Request object in that dictionary, so an error doesn't get returned when checking Request object parameters in inner_handler. But instead of mapping those 2 objects in that dictionary, we could probably just get rid of those dictionary mappings and do something like

`if any(key not in specificDep and specificDep.get(key) is not Request and specificDep.get(key) is not Response for key in paramList):'

and handle everything within that function.

sansyrox commented on robyn/__init__.py:
```diff
@@ -47,48 +48,52 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
-        This is base handler for all the route decorators
+        This is base handler for all the decorators
 
-        :param route_type str: route type between GET/POST/PUT/DELETE/PATCH/HEAD/OPTIONS/TRACE
```
Why delete these?

DarrenSeng replied:
They weren't deleted. Our fork was just outdated and didn't have any of the authentication changes + comments yet

sansyrox commented on robyn/__init__.py:
```diff
@@ -47,48 +48,52 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
+        :param route_type str: route type between GET/POST/PUT/DELETE/PATCH
         :param endpoint str: endpoint for the route added
         :param handler function: represents the sync or async function passed as a handler for the route
-        :param is_const bool: represents if the handler is a const function or not
-        :param auth_required bool: represents if the route needs authentication or not
```
And this?

DarrenSeng replied:
Our fork was outdated and didn't have any of the authentication changes + comments yet. Will fix that by the next commit.

sansyrox commented on robyn/__init__.py:
```diff
@@ -47,48 +48,52 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
 
         """ We will add the status code here only
         """
-        if auth_required:
-            self.middleware_router.add_auth_middleware(endpoint)(handler)
-
         return self.router.add_route(
-            route_type, endpoint, handler, is_const, self.exception_handler
```
Same

DarrenSeng replied:
I didn't sync our fork at the time the code for authentication was written. Will resolve in the next commit

sansyrox commented on robyn/__init__.py:
```diff
@@ -47,48 +48,52 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
+            route_type, endpoint, handler, is_const, self.dependencies,self.exception_handler #added deps
         )
+    
+    def inject(self, route = None, http_method=None, **kwargs:Callable[...,any]): 
```
I think we can go a bit lighter with the inline comments and sum them up in a docstring.

PEP8 - has certain suggestions about inline comments - https://peps.python.org/pep-0008/#inline-comments . It basically tells you to cautiously use the inline comments.

sansyrox commented on robyn/__init__.py:
```diff
@@ -99,7 +104,7 @@ def after_request(self, endpoint: Optional[str] = None) -> Callable[..., None]:
         """
         You can use the @app.after_request decorator to call a method after routing to the specified endpoint
 
-        :param endpoint str|None: endpoint to server the route. If None, the middleware will be applied to all the routes.
```
Is this required?

DarrenSeng replied:
I didn't write those comments. They are present in the original Robyn __init__.py too.

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
-            return self._add_route(
-                HttpMethod.GET, endpoint, handler, const, auth_required
-            )
+            return self._add_route(HttpMethod.GET, endpoint, handler, const) #when you declare function with get, will call _add_route
```
I think this is extra.

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
 
         return inner
 
-    def post(self, endpoint: str, auth_required: bool = False):
```
we should not remove this.

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
 
         def inner(handler):
-            return self._add_route(
-                HttpMethod.GET, endpoint, handler, const, auth_required
```
we should not remove this.

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
 
         def inner(handler):
-            return self._add_route(
-                HttpMethod.POST, endpoint, handler, auth_required=auth_required
```
same

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
 
         return inner
 
-    def put(self, endpoint: str, auth_required: bool = False):
```
we should not remove this.

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
 
         def inner(handler):
-            return self._add_route(
-                HttpMethod.TRACE, endpoint, handler, auth_required=auth_required
```
basically all the `auth_required` methods in this file.

DarrenSeng replied:
Yeah, our fork wasn't up to date when those changes were made. I'll fix that by the next commit

sansyrox commented on robyn/__init__.py:
```diff
@@ -218,133 +220,115 @@ def inner(handler):
         return inner
 
-    def include_router(self, router):
+    def include_router(self, router): #self = mainApp, router= sub_router
```
This comment is not needed imo

sansyrox commented on robyn/__init__.py:
```diff
@@ -365,22 +349,17 @@ def include_router(self, router):
+                continue
+            router.dependencies["all"][dep] = self.dependencies["all"][dep] #here, extend super's global deps to subrouters
 
-    def configure_authentication(self, authentication_handler: AuthenticationHandler):
-        """
-        Configures the authentication handler for the application.
-
-        :param authentication_handler: the instance of a class inheriting the AuthenticationHandler base class
-        """
-        self.authentication_handler = authentication_handler
-        self.middleware_router.set_authentication_handler(authentication_handler)
-
-
```
we need this.

sansyrox commented on robyn/router.py:
```diff
@@ -69,49 +75,76 @@ def _format_response(self, res):
+    
 
-
+#associated with middlewares. used for @app.before_request() and @app.after_request()
```
this too.

sansyrox commented on robyn/router.py:
```diff
@@ -69,49 +75,76 @@ def _format_response(self, res):
+            paramList = [a.name for a in argsFromHandler]
+            if any(key not in specificDep for key in paramList):
+                missing_deps = [key for key in paramList if key not in specificDep]
+                raise ValueError("Parameter not specified in dependencies dictionary:", missing_deps)
```
The error should be more like `Unknown Argument` as it is not necessary that the person wanted to refer to the dependency

DarrenSeng replied:
Will change it to `raise ValueError("Unknown Argument")`

sansyrox commented on robyn/router.py:
```diff
@@ -69,49 +75,76 @@ def _format_response(self, res):
             try:
-                response = self._format_response(await handler(*args))
+                if depToPass:
+                    response = self._format_response(await handler( *depToPass))#next(iter(depToPass.values()))))
```
do we need this?

DarrenSeng replied:
Do you mean the if/else statement? If so, then probably not. Could probably just be reduced to:
`try:`
`   response = self._format_response(await handler( *depToPass))`

but if you mean, "does it have to be *depToPass inside of await handler(...) instead of *args?", then yes. Unless you might have some suggestions for that.

sansyrox replied:
No, I was just talking about the comment 😅

DarrenSeng replied:
Ah ok, haha

sansyrox reviewed (commented):
Hey @DarrenSeng 👋 

Thank you for the PR 😄 I have left the first comment. Let me know if you require any more information from me. I will do another review after you folks address these 😄

sansyrox commented on robyn/__init__.py:
```diff
@@ -33,7 +33,7 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
         self.directory_path = directory_path
         self.config = config
-
+        self.dependencies = {"all":{'request':Request, 'response':Response}}
```
I have two suggestions here -

Let us replace `"all"` with `"ALL"` . And we should we are mapping the 'request' and 'response' keys to uninitialized objects. Why is that?

DarrenSeng replied:
The reason 'request' and 'response' are mapped to those objects is because when we construct the depsToPass in the inner_handler() functions,
`depToPass = [specificDep[key] for key in paramList if key in specificDep]   `
 you are considering every parameter specified for the handler function, and comparing it to the dependencies dictionary. If a parameter is present in that dictionary, then it gets added in depsToPass. However, request (not sure about response) objects might be a specified parameter for a function that still needs to be used as an arg when formatting the response at the line:
`response = self._format_response(await handler( *depToPass))`   
So by having 'request' specified in dependencies, depToPass construction will be able to account for a request parameter if the function was defined with one.

sansyrox replied:
@DarrenSeng , makes sense thank you. Can you also please put it as an inline comment?

DarrenSeng replied:
Added some comments for that

sansyrox commented on robyn/router.py:
```diff
@@ -76,12 +76,19 @@ def add_route(
     ) -> Union[Callable, CoroutineType]:
         @wraps(handler)
         async def async_inner_handler(*args):
+            argsFromHandler = (inspect.signature(handler)).parameters.values()
```
one suggestion for this file. We should be using `snake_case` in python. A lot of variables are declared in `camelCase`

DarrenSeng replied:
I just changed all the variables we made to snake_case

sansyrox commented:
Hey @DarrenSeng @koryca 👋 

Apologies for missing this PR. I will be giving a thorough review by tonight 😄

sansyrox commented on robyn/__init__.py:
```diff
@@ -33,7 +33,7 @@ def __init__(self, file_object: str, config: Config = Config()) -> None:
         self.directory_path = directory_path
         self.config = config
-
+        self.dependencies = {"ALL":{'request':Request, 'response':Response}}
```
Hey @DarrenSeng 👋 

Great progress so far 😄 

One more suggestion - can you extract this dictionary to a custom data structure?
```


i.e. like self.dependencies = DependencyMap()

class DependencyMap:
     """ This will have a dictionary inside"""
     
 ```
 
 This is called composition.
 
 A good read will be something like this - https://python-patterns.guide/gang-of-four/composition-over-inheritance/

DarrenSeng replied:
So something like?
`class DependencyMap:
    def __init__(self):
        self.dep_dict = {"ALL":{'request':Request, 'response':Response}}

class Robyn:
    def __init__(self, file_object: str, config: Config = Config()) -> None:
        self.dependencies = DependencyMap()
`

sansyrox replied:
@DarrenSeng , yes. And then the dependencymap will have certain accessor methods.

sansyrox commented:
Hey @DarrenSeng 👋 

I hope all is well 😄 

How are we looking here? Can we do anything to help with the PR?

DarrenSeng commented:
Hi 👋 I've committed an idea for the Dependency data structure. Unsure if it's a step in the right direction though

sansyrox commented on robyn/dependency_injection.py:
```diff
@@ -0,0 +1,55 @@
+        # this is a bit inneficient
+        # we are always adding the global dependencies to the route dependencies
+        # we should be able to to compose the basic data structure in a more efficient way
+        # cc @Darren
```
@DarrenSeng

DarrenSeng replied:
I addressed this logic in the most recent commit. Global dependencies no longer get constantly added to route dependencies. The issue I had before you brought this up was that I wasn't sure how handler functions associated with route specific dependencies could still make use of global "ALL_ROUTES" dependencies. In the commit, in router.py, I changed how deps_dict was constructed so that instead of 
`deps_dict = dependencies.get(endpoint, dependencies["ALL_ROUTES"])`
we have
`deps_dict = {**dependencies.get(endpoint, {}), **dependencies["ALL_ROUTES"]}`
This construction combines dependencies specific to the endpoint with the global "ALL" dependencies into a single dictionary deps_dict. If the endpoint is not specified, it will default to an empty dictionary for the endpoint-specific dependencies, effectively only using the global "ALL" dependencies.

sansyrox commented on robyn/__init__.py:
```diff
@@ -80,11 +90,30 @@ def _add_route(
 
+    def inject(self, route=None, http_method=None, **kwargs: dict):
+        # the http_method param is unused??
+        # @Darren
```
@DarrenSeng

DarrenSeng replied:
http_method as a param was shown in one of the example cases in one of your comments on the issue page, but I'm still not sure how to make use of it, so I'll probably delete that param

sansyrox commented on robyn/router.py:
still need to review this file

sansyrox reviewed (commented):
Hey @DarrenSeng 👋 

I have refactored the code in PR, and have also tagged you in a few comments. Do let me know if you have any questions for me 😄

IdoKendo commented on robyn/__init__.py:
```diff
@@ -367,6 +394,11 @@ def include_router(self, router):
+        for dep in self.dependencies.get_dependencies():
+            if dep in router.dependencies.get_dependencies():
+                continue
+            router.dependencies.get_dependencies()[dep] = self.dependencies.get_dependencies()[dep]
```
I think this piece of logic should sit in dependency_injection, and here just a call along the lines of:
```py
router.dependencies.extend(self.dependencies)
```

IdoKendo commented on robyn/router.py:
```diff
@@ -78,12 +79,28 @@ def add_route(
         async def async_inner_handler(*args):
+            handler_args = (inspect.signature(handler)).parameters.values()
+            param_list = [a.name for a in handler_args]
+            dependency_dict = {**dependencies.get(endpoint, {}), **dependencies["ALL_ROUTES"]}
```
Let's abstract this section outside of the handlers to a common logic function.

IdoKendo commented on robyn/dependency_injection.py:
```diff
@@ -0,0 +1,49 @@
+        #'request' and 'response' mappings are needed for when constructing deps_to_pass in router.py
+        self._dependency_map = {"ALL_ROUTES": {"request": Request, "response": Response}}
+
+    def add_route_dependency(self, route, **kwargs):
```
Can we add type hints for `route` in this file?

IdoKendo commented on robyn/dependency_injection.py:
```diff
@@ -0,0 +1,49 @@
+            route (str): The route to get the dependencies for.
+        """
+        if route:
+            if route in self._dependency_map:
```
There is a logic flaw here, if route is passed but it does not exist in `self._dependency_map` this will return None.

IdoKendo replied:
There is also discrepancy in the API for setting and getting dependencies, for setting there are 2 different functions, `set_route_dependency` and `set_global_dependency` but for getting there is one common function - I feel that this should be aligned, either way is fine IMO, but should be the same.

IdoKendo reviewed (changes requested):
Hey @DarrenSeng 👋  Great job!
I've added a few comments, let me know if something is unclear.
Also, would really appreciate if you could add some tests to validate the behavior as well as showing an example of how it's used.

sansyrox commented:
Thanks for the review @IdoKendo 😄

sansyrox commented:
Any updates on this @DarrenSeng ?

DarrenSeng commented:
@sansyrox Hello! Sorry for the delay in progress. I've addressed the comments from IdoKendo. Still need to create some tests to validate the behavior though

sansyrox commented on robyn/__init__.py:
```diff
@@ -100,9 +101,21 @@ def add_route(
+            return self.dependencies.get_route_dependencies(route)
+        else:
+            return self.dependencies.get_dependencies()
+
```
@DarrenSeng , `get_dependencies` is not a function anywhere.

sansyrox commented on robyn/dependency_injection.py:
```diff
@@ -0,0 +1,74 @@
+        This method iterates through the dependencies of this DependencyMap and adds any dependencies
+        that are not already present in the target router's DependencyMap. 
+        """
+        for dep_key in self.get_dependencies():
+            if dep_key in target_router.dependencies.get_dependencies():
+                continue
+            target_router.dependencies.get_dependencies()[dep_key] = self.get_dependencies()[dep_key] 
```
Same here. Where is this function defined?

sansyrox commented:
@DarrenSeng , I have added some sample tests. Also, you will have to rebase your branch to the latest main

sansyrox commented on robyn/__init__.py:
```diff
@@ -100,9 +101,25 @@ def add_route(
+
+    def get_injected_dependencies(self, route=None) -> dict:
+        # How will we access these from sub routes?
+        # @Darren, @Ido ?? Any ideas?
```
Hey @IdoKendo @DarrenSeng 👋 

I have one question regarding the public facing interface of the API. Do you folks have any suggestions?

sansyrox commented on main.py:
```diff
@@ -0,0 +1,63 @@
+from robyn import Robyn, SubRouter, jsonify
```
@DarrenSeng , this file should not exist. 

All this code should be a part of the `base_routes` files instead of what I wrote. 

And the corresponding routes should be tested the way demonstrated!

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -737,7 +739,30 @@ async def async_without_decorator():
+
+# ===== Dependency Injection =====
+#
+GLOBAL_DEPENDENCY = "global_dependency"
```
@DarrenSeng 👋 

Can we also show how this injection translates to Subrouters?

sansyrox commented:
Hey @DarrenSeng 👋 

I will be pushing some commits to this PR 😄

sansyrox commented on robyn/__init__.py:
```diff
@@ -407,6 +434,7 @@ def __init__(
         self, file_object: str, prefix: str = "", config: Config = Config()
     ) -> None:
         super().__init__(file_object, config)
+        dependencies = DependencyMap()
```
why is this not a part of the constructor?

sansyrox commented on robyn/router.py:
```diff
@@ -82,22 +83,46 @@ def _format_response(
             )
         return response
 
+    def validate_handler_args(self, handler, endpoint, dependencies):
```
we should be using this api to test. why are we using `app.get_injected_dependencies` to test?

sansyrox commented on robyn/router.py:
```diff
@@ -111,9 +136,25 @@ async def async_inner_handler(*args):
+
+            # dependencies_to_pass construction considers each parameter specified in the handler function
+            #'request' specified in init's dep mapping lets this construction account for a request parameter in the handler function
+            dependencies_to_pass = [
```
let us rename this to `injected_dependencies`

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -756,7 +757,68 @@ async def async_without_decorator():
+    return {"description": description}
+
+
+subrouter = SubRouter(__file__, "/subrouter")
+# in case of collision, sub router dependency will be given priority
+GLOBAL_DEP_COLLISION = "GLOBAL_DEP_COLLISION as defined in subrouter dependency dictionary"
+ROUTE_DEPENDENCY = "subrouter_route_dependency"
+subrouter.inject(GLOBAL_DEP_COLLISION=GLOBAL_DEP_COLLISION)
+subrouter.inject("/subrouter/dep", ROUTE_DEPENDENCY=ROUTE_DEPENDENCY)
+
+
+@subrouter.get("/subrouter_route_dep_inject")
+def sync_subrouter_route_dependency():
+    local_dependencies = subrouter.get_injected_dependencies("/subrouter/dep")
+    return {"description": local_dependencies["ROUTE_DEPENDENCY"]}
+
+
+@subrouter.get("/subrouter_global_dep_inject")
+def sync_subrouter_global_dependency():
+    global_dependencies = subrouter.get_injected_dependencies()
+    description = "\n".join(
+        [
+            f"GLOBAL_DEPENDENCY: {global_dependencies['GLOBAL_DEPENDENCY']}",
+            f"GLOBAL_DEP_COLLISION: {global_dependencies['GLOBAL_DEP_COLLISION']}",
+            f"GLOBAL_DEP_INHERIT: {global_dependencies['GLOBAL_DEP_INHERIT']}",
+        ]
+    )
+    return {"description": description}
```
This should be in the subrouter file only.

sansyrox commented on robyn/router.py:
```diff
@@ -77,22 +78,42 @@ def _format_response(
+            param_list, dependency_dict = self.validate_handler_args(handler, endpoint, dependencies)
+            # dependencies_to_pass construction considers each parameter specified in the handler function
+            #'request' specified in init's dep mapping lets this construction account for a request parameter in the handler function
+            dependencies_to_pass = [dependency_dict[key] for key in param_list if key in dependency_dict]
```
we need to include `*args` here.

sansyrox commented on robyn/__init__.py:
```diff
@@ -155,7 +175,7 @@ def _add_event_handler(self, event_type: Events, handler: Callable) -> None:
 
         is_async = asyncio.iscoroutinefunction(handler)
-        self.event_handlers[event_type] = FunctionInfo(handler, is_async, 0)
+        self.event_handlers[event_type] = FunctionInfo(handler, is_async, 0, {})
```
we need to make this available in middlewares too.

sansyrox commented on robyn/router.py:
```diff
@@ -174,6 +198,10 @@ def inner_handler(request: Request, *args):
     # They take a handler, convert it into a closure and return the arguments.
     # Arguments are returned as they could be modified by the middlewares.
     def add_middleware(self, middleware_type: MiddlewareType, endpoint: Optional[str]) -> Callable[..., None]:
+        global_dependency_map = {}
+        router_dependency_map = {}
```
fix this.

sansyrox commented:
Hey @DarrenSeng 👋 

The PR is complete now. This what I was expecting with dependency injections in Robyn.

@IdoKendo , I hope all is well. Would it be possible for you to give a review here? @DarrenSeng would you like to review too?

IdoKendo reviewed (approved):
Sorry for taking long to finish the review, but everything LGTM 🔥

sansyrox commented:
> Sorry for taking long to finish the review, but everything LGTM 🔥

Thank you for reviewing this long PR @IdoKendo 😄

sansyrox reviewed (approved):
LGTM! ✨
